### PR TITLE
Onnx load model

### LINKED ISSUE #5766: No way to specify temp file location.
Currently the user has no way of specifying the temp file location. This means for any temp files that ML.NET creates they are stored in the default temp file location. This is causing some problems for some users who either have small drives where the temp files go, or who want to make sure all resources are cleaned up correctly.

### PR DESCRIPTION
**Pull request by michaelgsharp**
Fixes #5766 and #5764 by letting the user specify a temp path location, which defaults to the current location to keep existing behavior, and by make sure the temp onnx model file gets cleaned up by opening it with a delete on close flag.

### CODE REVIEW

File: src/Microsoft.ML.Data/MLContext.cs [hunk 1]
```diff
@@ -79,6 +79,13 @@ public sealed class MLContext : ISeededEnvironment
         /// </summary>
         public ComponentCatalog ComponentCatalog => _env.ComponentCatalog;
 
+        private static string _tempFilePath = System.IO.Path.GetTempPath();
+        public static string TempFilePath
+        {
+            get { return _tempFilePath; }
+            set { _tempFilePath = value; }
+        }
```
Comment (tarekgh) [inline src/Microsoft.ML.Data/MLContext.cs:87]: nit: you can write the whole things here as:
```C#
public static string TempFilePath {get; set; } = System.IO.Path.GetTempPath();
```

File: src/Microsoft.ML.Data/MLContext.cs [hunk 2]
```diff
@@ -79,6 +79,13 @@ public sealed class MLContext : ISeededEnvironment
         /// </summary>
         public ComponentCatalog ComponentCatalog => _env.ComponentCatalog;
 
+        private static string _tempFilePath = System.IO.Path.GetTempPath();
```
Comment (tarekgh) [inline src/Microsoft.ML.Data/MLContext.cs:82]: >System.IO [](http://example.com/codeflow?start=46&length=9)

does it make sense to add "using System.IO" in the top of the file?

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs [hunk 3]
```diff
@@ -202,9 +195,15 @@ public OnnxVariableInfo(string name, OnnxShape shape, Type typeInOnnxRuntime, Da
             {
                 // Load ONNX model file and parse its input and output schema. The reason of doing so is that ONNXRuntime
                 // doesn't expose full type information via its C# APIs.
-                ModelFile = modelFile;
                 var model = new OnnxCSharpToProtoWrapper.ModelProto();
-                using (var modelStream = File.OpenRead(modelFile))
+                // If we own the model file set the DeleteOnClose flag so it is always deleted.
+                if (ownModelFile)
```
Comment (tarekgh) [inline src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:200]: >if [](http://example.com/codeflow?start=16&length=2)

nit: please use parentheses with the if statements.

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs [hunk 4]
```diff
@@ -348,7 +347,7 @@ public static OnnxModel CreateFromBytes(byte[] modelBytes)
         public static OnnxModel CreateFromBytes(byte[] modelBytes, int? gpuDeviceId = null, bool fallbackToCpu = false,
             IDictionary<string, int[]> shapeDictionary = null)
         {
-            var tempModelFile = Path.GetTempFileName();
+            var tempModelFile = Path.Combine(MLContext.TempFilePath, Path.GetRandomFileName());
```
Comment (tarekgh) [inline src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:350]: >Path.Combine(MLContext.TempFilePath, Path.GetRandomFileName()); [](http://example.com/codeflow?start=32&length=63)

I am seeing this is repeated. would make sense to have a property in MLContext to return the complete random file path?

**Review by tarekgh — APPROVED** [review]
Added some minor comments. otherwise, LGTM.

File: src/Microsoft.ML.Data/MLContext.cs [hunk 5]
```diff
@@ -79,6 +79,13 @@ public sealed class MLContext : ISeededEnvironment
...
+        public static string TempFilePath
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/MLContext.cs:83]: Add xml documentation to the new API

Comment (eerhardt) [inline src/Microsoft.ML.Data/MLContext.cs:83]: This shouldn't be `static`. It should be an instance property on MLContext.

File: src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs [hunk 6]
```diff
@@ -304,7 +304,7 @@ private protected override void SaveModel(ModelSaveContext ctx)
             ctx.CheckAtModel();
             ctx.SetVersionInfo(GetVersionInfo());
 
-            ctx.SaveBinaryStream("OnnxModel", w => { w.WriteByteArray(File.ReadAllBytes(Model.ModelFile)); });
+            ctx.SaveBinaryStream("OnnxModel", w => { w.WriteByteArray(File.ReadAllBytes(Model.ModelStream.Name)); });
```
Comment (eerhardt) [inline src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs:307]: Is this change intentional?

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs (hunk 4 above)
Comment (eerhardt) [reply src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:350]: If we add this, it should be `internal`. We don't want to expose it publicly.

File: test/Microsoft.ML.PerformanceTests/FeaturizeTextBench.cs [hunk 7]
```diff
@@ -25,9 +25,8 @@ public class FeaturizeTextBench : BenchmarkBase
         [GlobalSetup]
         public void SetupData()
         {
-            Path.GetTempFileName();
```
Comment (eerhardt) [inline test/Microsoft.ML.PerformanceTests/FeaturizeTextBench.cs:28]: Do any of these tests need to change? I think it would be fine doing what they do today.

**Review by eerhardt — DISMISSED** [review]
The new property shouldn't be static. We should change it to an instance property.

File: src/Microsoft.ML.Data/MLContext.cs (hunk 1 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Data/MLContext.cs:87]: That is what I did originally, but ML.NET has a "style cop" and it complains if I do that. I can look into disabling it though for that line.

File: src/Microsoft.ML.Data/MLContext.cs (hunk 5 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Data/MLContext.cs:83]: The reason I made it static is there are places we don't pass the ML Context to that use a temp path. Do you have some time to sync on this today?

File: src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs (hunk 6 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.OnnxTransformer/OnnxTransform.cs:307]: Yeah, Before we just tracked the file name, but if the user didn't dispose of the OnnxTransformer the file would stay on disk. Now, we are keeping a file handle with `Delete on close` set, so this change is just to get the name from the handle instead.

File: test/Microsoft.ML.PerformanceTests/FeaturizeTextBench.cs (hunk 7 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.PerformanceTests/FeaturizeTextBench.cs:28]: Sounds good. I will go ahead and revert them then.

File: src/Microsoft.ML.Data/MLContext.cs (hunk 2 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Data/MLContext.cs:82]: I didn't just because I am only using it in that one place. But I'm happy to switch it.

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs (hunk 4 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:350]: Its a new random file path each time, so we probably wouldn't want it to be a property. I could add an internal method that returned that though.

File: src/Microsoft.ML.AutoML/API/ExperimentSettings.cs [hunk 8]
```diff
@@ -66,7 +66,7 @@ public ExperimentSettings()
         {
             MaxExperimentTimeInSeconds = 24 * 60 * 60;
             CancellationToken = default;
-            CacheDirectory = new DirectoryInfo(Path.Combine(MLContext.TempFilePath, "Microsoft.ML.AutoML"));
+            CacheDirectory = new DirectoryInfo(Path.Combine(Path.GetTempPath(), "Microsoft.ML.AutoML"));
```
Comment (eerhardt) [inline src/Microsoft.ML.AutoML/API/ExperimentSettings.cs]: Doesn't `ExperimentSettings.cs ` still need to get the TempFilePath here?

File: src/Microsoft.ML.Data/Commands/TrainTestCommand.cs [hunk 9]
```diff
@@ -188,7 +188,7 @@ private void RunCore(IChannel ch, string cmd)
 
             ILegacyDataLoader testPipe;
             bool hasOutfile = !string.IsNullOrEmpty(ImplOptions.OutputModelFile);
-            var tempFilePath = hasOutfile ? null : Path.GetTempFileName();
+            var tempFilePath = hasOutfile ? null : Path.Combine((Host as IHostEnvironmentInternal).TempFilePath, Path.GetRandomFileName());
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Commands/TrainTestCommand.cs:191]: Don't use `as` in these places. Instead just straight cast. That way you get an InvalidCastException when it is wrong, instead of a NullReferenceException.

File: src/Microsoft.ML.Data/MLContext.cs [hunk 10]
```diff
@@ -79,6 +79,15 @@ public sealed class MLContext : ISeededEnvironment
         /// </summary>
         public ComponentCatalog ComponentCatalog => _env.ComponentCatalog;
 
+        /// <summary>
+        /// The location for the temp files created by ML.NET
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/MLContext.cs:83]:
```
Gets or sets the location for the temp files created by ML.NET
```

File: src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs [hunk 11]
```diff
@@ -377,6 +380,7 @@ public void RemoveListener(Action<IMessageSource, TMessage> listenerFunc)
             string shortName = null, string parentFullName = null)
             : base(shortName, parentFullName, verbose)
         {
+            TempFilePath = Path.GetTempPath();
```
Comment (eerhardt) [inline src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs:383]: This should read it from the `source` environment.

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs [hunk 12]
```diff
@@ -143,14 +143,9 @@ public OnnxVariableInfo(string name, OnnxShape shape, Type typeInOnnxRuntime, Da
         /// </summary>
         private readonly InferenceSession _session;
         /// <summary>
-        /// Indicates if <see cref="ModelFile"/> is a temporal file created by <see cref="CreateFromBytes(byte[], int?, bool, IDictionary{string, int[]})"/>
-        /// or <see cref="CreateFromBytes(byte[])"/>. If <see langword="true"/>, <see cref="Dispose(bool)"/> should delete <see cref="ModelFile"/>.
-        /// </summary>
-        private bool _ownModelFile;
-        /// <summary>
         /// The location where the used ONNX model loaded from.
```
Comment (eerhardt) [inline src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:146]: This comment is now out of date.

File: src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs [hunk 13]
```diff
@@ -93,9 +93,11 @@ internal interface IMessageSource
     /// query progress.
     /// </summary>
     [BestFriend]
-    internal abstract class HostEnvironmentBase<TEnv> : ChannelProviderBase, ISeededEnvironment, IChannelProvider, ICancelable
+    internal abstract class HostEnvironmentBase<TEnv> : ChannelProviderBase, IHostEnvironmentInternal, IChannelProvider, ICancelable
         where TEnv : HostEnvironmentBase<TEnv>
     {
+        public string TempFilePath { get; set; }
```
Comment (eerhardt) [inline src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs:99]: This should go with the rest of the properties on this class. See lines 331-356

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs (hunk 3 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:200]: This is the style that most of ML.NET, and this file, use for doing single line if statements. To keep things consistent I am going to leave it as is.

File: src/Microsoft.ML.Data/MLContext.cs (hunk 5 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Data/MLContext.cs]: Added.

File: test/Microsoft.ML.PerformanceTests/FeaturizeTextBench.cs (hunk 7 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.PerformanceTests/FeaturizeTextBench.cs:28]: Tests have been reverted.

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs (hunk 12 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:146]: Comment has been removed.

File: src/Microsoft.ML.AutoML/API/ExperimentSettings.cs [hunk 14]
```diff
@@ -70,6 +70,18 @@ public ExperimentSettings()
             CacheBeforeTrainer = CacheBeforeTrainer.Auto;
             MaxModels = int.MaxValue;
         }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ExperimentSettings"/> using the temp file location provided to the MLContext.
+        /// </summary>
+        public ExperimentSettings(MLContext context)
```
Comment (eerhardt) [inline src/Microsoft.ML.AutoML/API/ExperimentSettings.cs:77]: I'm not really excited about this now that I see it. I didn't understand the issue previously. Mostly, I don't like the duplication, and how many things needed to change.

Note that AutoML is still an "experimental" library, which means we can make breaking changes to the API, if we believe they are the right changes going forward.

I can think of a couple alternatives:

1. Don't make any changes to AutoML - keep the `CacheDirectory` property defaulting to `Path.GetTempPath()`, and not respecting `MLContext.TempFilePath`. The user has full control, but they just need to set 2 different properties.
2. Remove `CacheDirectory` from ExperiementSettings, and just use `context.TempFilePath` where `CacheDirectory` was used before.
2.b. Rename `CacheDirectory` to just be the directory name (i.e. `CacheDirectoryName`) and create that directory under `MLContext.TempFilePath`.

Thoughts?

Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/API/ExperimentSettings.cs:77]: I think option 2b is the best. AutoML is just supposed to help figure out pipelines automatically, I think it still makes sense to keep a setting like that inside of the MLContext itself. I do think having the option of a separate folder though is good because if that value is set then AutoML will write all the models to disk, so it makes sense to have a folder just for them.

Comment (eerhardt) [reply src/Microsoft.ML.AutoML/API/ExperimentSettings.cs:77]: That sounds good to me. Thanks!

Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/API/ExperimentSettings.cs]: Alright, change 2b has been implemented and pushed. Take a look and let me know if its what you were thinking.

File: src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs [hunk 15]
```diff
@@ -6,6 +6,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using Microsoft.ML.Data;
+using Microsoft.ML.Runtime;
```
Comment (eerhardt) [inline src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:9]: Can this be reverted?

File: src/Microsoft.ML.AutoML/API/ExperimentSettings.cs [hunk 16]
```diff
@@ -60,16 +60,17 @@ public abstract class ExperimentSettings
         internal int MaxModels;
 
         /// <summary>
-        /// Initializes a new instance of <see cref="ExperimentSettings"/>.
+        /// Initializes a new instance of <see cref="ExperimentSettings"/> using the default temp file location.
```
Comment (eerhardt) [inline src/Microsoft.ML.AutoML/API/ExperimentSettings.cs:63]:
```suggestion
        /// Initializes a new instance of <see cref="ExperimentSettings"/>.
```

File: src/Microsoft.ML.AutoML/Experiment/Experiment.cs [hunk 17]
```diff
@@ -56,7 +56,7 @@ internal class Experiment<TRunDetail, TMetrics> where TRunDetail : RunDetail
             _experimentSettings = experimentSettings;
             _metricsAgent = metricsAgent;
             _trainerAllowList = trainerAllowList;
-            _modelDirectory = GetModelDirectory(_experimentSettings.CacheDirectory);
+            _modelDirectory = GetModelDirectory(new DirectoryInfo(Path.Combine(_context.TempFilePath,_experimentSettings.CacheDirectoryName)));
```
Comment (eerhardt) [inline src/Microsoft.ML.AutoML/Experiment/Experiment.cs:59]: This isn't going to work with the intended behavior from the XML comments on `CacheDirectoryName`:

> If <see langword="null"/>, models will be kept in memory instead of written to disk.

You will need to pass both `_context` and `_experimentSettings.CacheDirectoryName` into this method, and check `CacheDirectoryName` for null, and return `null`.

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs (hunk 12 above)
Comment (eerhardt) [reply src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:146]: The comment still isn't removed.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs [hunk 18]
```diff
@@ -1322,9 +1323,9 @@ private void VariableSummaries(ResourceVariable var)
         private static TensorFlowSessionWrapper LoadTensorFlowSessionFromMetaGraph(IHostEnvironment env, Architecture arch)
         {
             var modelFileName = ModelFileName[arch];
-            var modelFilePath = Path.Combine(_resourcePath, modelFileName);
+            var modelFilePath = Path.Combine((env as IHostEnvironmentInternal).TempFilePath, "MLNET", modelFileName);
```
Comment (eerhardt) [inline src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1326]: It maybe better to make this an instance method and continue to use `_resourcePath`.

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 19]
```diff
@@ -183,16 +183,16 @@ public void AutoFitRegressionTest(string culture)
                 // Furthermore, these issues might only occur after ~70
                 // iterations, so setting the internal maxModels parameter.
                 int maxModels = culture == "en-US" ? 1 : 75;
+                var context = new MLContext(1);
```
Comment (eerhardt) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:186]: Can these test changes (and the ones in the next file) be reverted? Looks like they aren't necessary anymore.

**Review by eerhardt — COMMENTED** [review]
Yes, this is the direction I had in mind. A few more comments to fix, then after that I think this will be ready to be merged.

File: src/Microsoft.ML.AutoML/Experiment/Experiment.cs (hunk 17 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/Experiment/Experiment.cs:59]: Ah, good catch. Thanks, I'll get that fixed!

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs (hunk 19 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:186]: I believe so. I'll revert them and test it.

File: src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs (hunk 15 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:9]: Yep. I'll revert it.

File: src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs (hunk 12 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.OnnxTransformer/OnnxUtils.cs:146]: Ok, now it says `The FileStream holding onto the loaded ONNX model.`. It will go with my next push.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs (hunk 18 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1326]: I tried that and it caused other issues. Let me make that change again and I'll give more details. IIRC though it has to do with us parsing the graph before we have instantiated it here.

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs (hunk 19 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:186]: Reverted.

File: src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs (hunk 15 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:9]: Reverted

File: src/Microsoft.ML.AutoML/Experiment/Experiment.cs (hunk 17 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/Experiment/Experiment.cs]: Ok, fixed.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs (hunk 18 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1326]: Nevermind, I was thinking of something else. Changed it to an instance method and reverted to `_resourcePath`.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs [hunk 20]
```diff
@@ -1319,12 +1320,12 @@ private void VariableSummaries(ResourceVariable var)
 
         }
 
-        private static TensorFlowSessionWrapper LoadTensorFlowSessionFromMetaGraph(IHostEnvironment env, Architecture arch)
+        private TensorFlowSessionWrapper LoadTensorFlowSessionFromMetaGraph(IHostEnvironment env, Architecture arch)
         {
             var modelFileName = ModelFileName[arch];
-            var modelFilePath = Path.Combine(_resourcePath, modelFileName);
+            var modelFilePath = Path.Combine(_resourcePath, "MLNET", modelFileName);
```
Comment (eerhardt) [inline src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1326]: >M [](http://example.com/codeflow?start=61&length=1)

Doesn't `_resourcePath` already have `MLNET` appended to it, from the constructor?

Comment (michaelgsharp) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs]: It did yeah, and it is what was failing the tests. That has been resolved now.

**Review by eerhardt — APPROVED** [review]
LGTM. Nice work here.
